Allow optimization of PBS select statements

milljm opened the pull request:
Add ability for RunPBS to better calculate the select statement.

Modify util.runCommand so it can handle large amounts of output.
In particular, `pbsnodes -a` can possibly generate enough output
to break current design:
(https://docs.python.org/2/library/subprocess.html#popen-objects).

Closes #11609

Commit:
Allow optimization of PBS select statements

Add ability for RunPBS to better calculate the select statement.

Modify util.runCommand so it can handle large amounts of output.
In particular, `pbsnodes -a` can possibly generate enough output
to break current design:
(https://docs.python.org/2/library/subprocess.html#popen-objects).

Closes #11609

Commit:
fix pyFlakes issues

missing import

Closes #11609

Commit:
refactor some logic and type settings

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/197416/) on 582c426 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11616/site)

This comment will be updated on new commits.

permcody commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -70,15 +90,37 @@ def hasTimedOutOrFailed(self, job_data):
+
+        # slots = procs_per_node
+        else:
+            select_ncpus = 'ncpus=%d' % (procs_per_node)
```
This logic is definitely not correct. There are indeed three cases but these aren't it.

1) both num_full_chunks and remaining ranks are nonzero (slots >= procs_per_node and not divisible by procs_per_node)
2) num_full_chunks is zero (slots < procs_per_node)
3) num_remaining_ranks is zero (slots >= procs_per_node but evenly divisable)

This is absolutely not working right now.

milljm replied:
Hmmm... then I don't understand. This totally creates the right qsub select statement for me.

milljm replied:
example of a 1K core run on rod with 32 cores available per node:
```pre
#PBS -l select=1:ncpus=8+31:ncpus=32
```
8 + (31 * 32) = 1000
remaining_ranks + (num_full_chunks * procs_per_node)

I need help understanding where I am going wrong with this.

Commit:
Fix the broken select statement

The select statement was building the node count very very
incorrectly :-)

Skip the job, if somehow someone over-subscribes on their
cluster.

permcody commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -70,15 +89,46 @@ def hasTimedOutOrFailed(self, job_data):
+
+        # slots > procs_per_node and is not evenly distributable (both non-zero)
+        if num_full_chunks and remaining_ranks:
+            select_ncpus = 'select=%d:ncpus=%d+%d' % (num_full_chunks, procs_per_node, remaining_ranks)
```
This still isn't right. The remaining ranks bit should be like the case below, not just "+n". I think a better way to generate this is to generate a "left and right" part and put them together when they are both non-zero with the plus otherwise just use one or the other halves.

milljm replied:
I am still confused... Let me be literal for a moment:

```python
        # get the common processor per node count
        left = int(max(groupby(sorted(all_nodes)), key=lambda(x, v):(len(list(v)),-all_nodes.index(x)))[0])

        chunks = slots / left
        right = slots % left

        # left is _always_ true
        if True and right:
            select = 'ncpus=%d+%d' % (left, right)

        else:
            select = 'ncpus=%d' % (left)

        if chunks:
            select = 'select=%d:%s' % (chunks, select)

        else:
            select = 'select=1:%s' % (select)
```
What is this 'left' part/integer you are talking about? Where does it come from?

permcody replied:
What I meant by "left and right" was the parts before and after the plus sign. More specifically, this is the "whole" and "partial" nodes.

milljm replied:
Hmmm, well that is still confusing because 'left' or 'whole' is always true (Its our processor count per node).

It is 'chunks' and 'partial' that can either be true or false... And if we have a _chunk_, then the left side of ncpus will _always_ be processor count per node.

And so, that is why I am doing the following (not caring what the left side of ncpus might be):
```python

        whole = slots / procs_per_node
        partial = slots % procs_per_node

        if whole and partial:
            select = 'select=whole:ncpus=procs_per_node+partial'
        elif partial:
            select = 'select=1:ncpus=partial'
        else:
            select = 'select=whole:npucs=procs_per_node'
```

We can do it the other way, but it takes more lines of code:

```python
        whole = slots / procs_per_node
        partial = slots % procs_per_node

        if procs_per_node and partial:
            ncpus = 'ncpus=procs_per_node+partial'
        else:
            ncpus = 'ncpus=procs_per_node'

        if whole:
            select = 'select=whole:ncpus'
        else:
            select = 'select=1:ncpus'
```

permcody commented:
On Mon, Jun 18, 2018 at 8:55 AM Jason Miller <notifications@github.com>
wrote:

> *@milljm* commented on this pull request.
> ------------------------------
>
> In python/TestHarness/schedulers/RunPBS.py
> <https://github.com/idaholab/moose/pull/11616#discussion_r196109026>:
>
> > +        all_nodes = self._getPBSResources()
> +        total_cores = sum([int(cores) for cores in all_nodes])
> +
> +        # Is this job to big for this cluster?
> +        if slots > total_cores:
> +            job.addCaveats('insufficient slots')
> +            job.setStatus(job.skip)
> +            # continue and allow the qsub script to be created
> +
> +        procs_per_node = int(max(groupby(sorted(all_nodes)), key=lambda(x, v):(len(list(v)),-all_nodes.index(x)))[0])
> +        num_full_chunks = slots / procs_per_node
> +        remaining_ranks = slots % procs_per_node
> +
> +        # slots > procs_per_node and is not evenly distributable (both non-zero)
> +        if num_full_chunks and remaining_ranks:
> +            select_ncpus = 'select=%d:ncpus=%d+%d' % (num_full_chunks, procs_per_node, remaining_ranks)
>
> Hmmm, well that is still confusing because 'left' or 'whole' is always
> true (Its our processor count per node).
>

No - If I request 32 processors I would have zero full nodes and one
partial one. Either can be true.


> It is 'chunks' and 'partial' that can either be true or false... And if we
> have a *chunk*, then the left side of ncpus will *always* be processor
> count per node.
>
> And so, that is why I am doing the following (not caring what the left
> side of ncpus might be):
>
>         whole = slots / procs_per_node
>         partial = slots % procs_per_node
>
>         if whole and partial:
>             select = 'select=whole:ncpus=procs_per_node+partial'
>         elif partial:
>             select = 'select=1:ncpus=partial'
>         else:
>             select = 'select=whole:npucs=procs_per_node'
>
> We can do it the other way, but it takes more lines of code:
>
>         whole = slots / procs_per_node
>         partial = slots % procs_per_node
>
>         if procs_per_node and partial:
>             ncpus = 'ncpus=procs_per_node+partial'
>         else:
>             ncpus = 'ncpus=procs_per_node'
>
>         if whole:
>             select = 'select=whole:ncpus'
>         else:
>             select = 'select=1:ncpus'
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/11616#discussion_r196109026>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIMZSx9H587u-IMbkA8bKWGWMDKUPks5t979NgaJpZM4UVx49>
> .
>
